util/ivmFactory.js

```diff
@@ -38,31 +38,64 @@ async function createIvm(code, libraryVersionIds) {
       const eventMessages = events.map(event => event.message);
       const eventsMetadata = {};
       events.forEach(ev => {
-        eventsMetadata[ev.message.messageId] = ev.metadata;
+        eventsMetadata[ev.metadata ? ev.metadata.messageId : ev.message.messageId] = ev.metadata;
       });
 
+      const isObject = (o) => Object.prototype.toString.call(o) === '[object Object]';
+
       var metadata = function(event) {
         const eventMetadata = event ? eventsMetadata[event.messageId] || {} : {};
-        return eventMetadata
+        return eventMetadata;
       }
       switch(transformType) {
         case "transformBatch":
-          outputEvents = await transformBatch(eventMessages, metadata);
-          outputEvents = outputEvents.map(transformedEvent => ({transformedEvent, metadata: metadata(transformedEvent)}))
+          const transformedEventsBatch = await transformBatch(eventMessages, metadata);
+          if (!Array.isArray(transformedEventsBatch)) {
+            outputEvents.push({error: "returned events from transformBatch(event) is not an array", metadata: {}});
+            break;
+          }
+          outputEvents = transformedEventsBatch.map(transformedEvent => {
+            if (!isObject(transformedEvent)) {
+              return{error: "returned event in events array from transformBatch(events) is not an object", metadata: {}};
+            }
+            return{transformedEvent, metadata: metadata(transformedEvent)};
+          })
           break;
         case "transformEvent":
-
-          outputEvents = await Promise.all(eventMessages.map(async ev => {
+          await Promise.all(eventMessages.map(async ev => {
+            const currMsgId = ev.messageId;
             try{
-              const currMsgId = ev.messageId;
-              let transformedEvent = await transformEvent(ev, metadata);
-              return {transformedEvent, metadata: eventsMetadata[currMsgId] || {}};
+              let transformedOutput = await transformEvent(ev, metadata);
+              // if func returns null/undefined drop event
+              if (transformedOutput === null || transformedOutput === undefined) return;
+              if (Array.isArray(transformedOutput)) {
+                const producedEvents = [];
+                let encounteredError = false;
+                transformedOutput.every(e => {
```

Comment:
Why use `.every` when you don't collect it's return?

Reply:
to break when non object is found